### PR TITLE
LTD limits

### PR DESCRIPTION
**Pull request by szerintedmi**
calculate and enforce LTD limits based on ltdParams:

- lockDifferenceLimit
- loanDifferenceLimit
- allowedDifferenceAmount

### CODE REVIEW

File: contracts/MonetarySupervisor.sol [hunk 1]
```diff
@@ -37,29 +39,31 @@ contract MonetarySupervisor is Restricted, TokenReceiver { // solhint-disable-li
 
     /* Parameters Used to ensure totalLoanAmount or totalLockedAmount difference is withing limit and system also works
         when any of those 0 or low. */
-    uint public ltdDifferenceLimit;     /* allow lock or loan if Loan To Deposut ratio stay within 1 +- this param
-                                            stored as parts per million */
+    uint public ltdLockDifferenceLimit;     /* allow lock or loan if Loan To Deposut ratio stay within 1 +- this param.
+                                            Stored as parts per million */
+    uint public ltdLoanDifferenceLimit;
     uint public allowedLtdDifferenceAmount; /* in token - if totalLoan and totalLock difference is less than that
                                              then allow loan or lock even if ltdDifference limit would go off with it */
```
Comment (phraktle) [inline contracts/MonetarySupervisor.sol]: please clarify this comment, it's unclear to me.

File: contracts/MonetarySupervisor.sol [hunk 2]
```diff
@@ -140,8 +149,54 @@ contract MonetarySupervisor is Restricted, TokenReceiver { // solhint-disable-li
     }
 
     // helper function for FrontEnd to reduce calls
-    function getParams() external view returns(uint[2]) {
-        return [ltdDifferenceLimit, allowedLtdDifferenceAmount];
+    function getParams() external view returns(uint[3]) {
+        return [ltdLockDifferenceLimit, ltdLoanDifferenceLimit, allowedLtdDifferenceAmount];
+    }
+
+    function getLoanToDepositRatio() external view returns (uint loanToDepositRatio) {
```
[the remark this answers is not in the document]
Comment (szerintedmi) [reply contracts/MonetarySupervisor.sol]: good call. LTD seems to be the jargon & would be a too big refactor to change name everywhere. I would leave this in the backlog

File: contracts/MonetarySupervisor.sol [hunk 3]
```diff
@@ -37,29 +39,31 @@ contract MonetarySupervisor is Restricted, TokenReceiver { // solhint-disable-li
 
     /* Parameters Used to ensure totalLoanAmount or totalLockedAmount difference is withing limit and system also works
         when any of those 0 or low. */
-    uint public ltdDifferenceLimit;     /* allow lock or loan if Loan To Deposut ratio stay within 1 +- this param
-                                            stored as parts per million */
+    uint public ltdLockDifferenceLimit;     /* allow lock or loan if Loan To Deposut ratio stay within 1 +- this param.
```
[the remark this answers is not in the document]
Comment (szerintedmi) [reply contracts/MonetarySupervisor.sol]: `lockLtdTolerance` ? `newLockLtdTolerance` ?
I'm tempted to leave this change in the backlog as this is complex anyways, not sure if a renaming would help too much..

File: contracts/MonetarySupervisor.sol [hunk 4]
```diff
@@ -140,8 +149,54 @@ contract MonetarySupervisor is Restricted, TokenReceiver { // solhint-disable-li
...
+
+        uint allowedByLtdDifferenceAmount = totalLoanAmount > totalLockedAmount.add(allowedLtdDifferenceAmount) ?
+                                        0 : totalLoanAmount.add(allowedLtdDifferenceAmount).sub(totalLockedAmount);
+
+        maxLockByLtd = allowedByLtdDifferencePt > allowedByLtdDifferenceAmount ?
+                                        allowedByLtdDifferencePt : allowedByLtdDifferenceAmount;
+    }
+
+    /* returns maximum borrowable token amount allowed by LTD params */
+    function getMaxLoanAmountAllowedByLtd() public view returns(uint maxLoanByLtd) {
```
Comment (phraktle) [inline contracts/MonetarySupervisor.sol]: `getMaxLoan*` and `getMaxLock*` are basically identical. refactor to a single function.

File: contracts/MonetarySupervisor.sol [hunk 5]
```diff
@@ -140,8 +149,54 @@ contract MonetarySupervisor is Restricted, TokenReceiver { // solhint-disable-li
     }
 
     // helper function for FrontEnd to reduce calls
-    function getParams() external view returns(uint[2]) {
-        return [ltdDifferenceLimit, allowedLtdDifferenceAmount];
+    function getParams() external view returns(uint[3]) {
```
Comment (phraktle) [inline contracts/MonetarySupervisor.sol]: LTD related settings should be in a single struct (eg. `LtdParams`), making it simpler to get/set and avoiding the need to prefix the fields.

File: contracts/MonetarySupervisor.sol (hunk 4 above)
Comment (szerintedmi) [reply contracts/MonetarySupervisor.sol]: not sure I got you.
getMaxLockAmount is an 'extension' of getMaxLockAmountAllowedByLtd 
getMaxLockAmount is a helper function for UI (when selected lock product is known)
getMaxLockAmountAllowedByLtd is used internally by requestInterest() .
we could make internal fx (minor benefits) but unit tests are testing them at the moment.

Comment (phraktle) [reply contracts/MonetarySupervisor.sol]: the point is that there's basically copy-pasted code shared by the two methods. you should extract this part into a common method. something along these lines:

```
    function max(uint allowedByLtdDifferencePt) internal returns(uint max) {
        allowedByLtdDifferencePt = totalLockedAmount >= allowedByLtdDifferencePt ?
                                        0 : allowedByLtdDifferencePt.sub(totalLockedAmount);

        uint allowedByLtdDifferenceAmount = totalLoanAmount > totalLockedAmount.add(allowedLtdDifferenceAmount) ?
                                        0 : totalLoanAmount.add(allowedLtdDifferenceAmount).sub(totalLockedAmount);

        return allowedByLtdDifferencePt > allowedByLtdDifferenceAmount ?
                                        allowedByLtdDifferencePt : allowedByLtdDifferenceAmount;

    }
```

Comment (szerintedmi) [reply contracts/MonetarySupervisor.sol]: ok got you. but the first line is not the same.  But indeed, this part can be moved to a generic fx:
```js
uint allowedByLtdDifferenceAmount = totalLoanAmount > totalLockedAmount.add(allowedLtdDifferenceAmount) ?
            0 : totalLockedAmount.add(allowedLtdDifferenceAmount).sub(totalLoanAmount);

maxLoanByLtd = allowedByLtdDifferencePt > allowedByLtdDifferenceAmount ?
             allowedByLtdDifferencePt : allowedByLtdDifferenceAmount;
```

Comment (szerintedmi) [reply contracts/MonetarySupervisor.sol]: well even that part is not the same :) 
can't figure a concise & readable way to make it a common function. Leaving it as unless you have a better idea

**Pull request merged**